src/lib.rs

```diff
@@ -156,24 +131,29 @@ where
     }
 
     /// Sets the single pin mode to either Mode::INPUT or Mode::OUTPUT
-    pub fn pin_mode(&mut self, pin: u8, pin_mode: PinMode) -> Result<(), E>  {
-        self.update_register_bit(pin, pm(pin_mode), Register::IODIRA, Register::IODIRB)
+    pub fn pin_mode(&mut self, pin: u8, pin_mode: PinMode) -> Result<(), E> {
+        self.update_register_bit(
+            pin,
+            pin_mode.bit_value(),
+            Register::IODIRA,
+            Register::IODIRB,
+        )
     }
 
     /// Sets pin mode to either Mode::INPUT or Mode::OUTPUT
-    pub fn all_pin_mode(&mut self, pin_mode: PinMode) -> Result<(), E>  {
-        self.write_register(Register::IODIRA, pm(pin_mode))?;
-	    self.write_register(Register::IODIRB, pm(pin_mode))
+    pub fn all_pin_mode(&mut self, pin_mode: PinMode) -> Result<(), E> {
+        self.write_register(Register::IODIRA, pin_mode.register_value())?;
+        self.write_register(Register::IODIRB, pin_mode.register_value())
     }
 
     /// Reads all 16 pins (port A and B) into a single 16 bits variable
-    pub fn read_gpioab(&mut self) -> Result<u16, E>  {
-        let buffer: GenericArray<u8, U2> = self.read_registers(Register::GPIOA)?;
-        Ok((u16(buffer[0]) << 8) + u16(buffer[1]))
+    pub fn read_gpioab(&mut self) -> Result<u16, E> {
+        let buffer = self.read_double_register(Register::GPIOA)?;
+        Ok((buffer[0] as u16) << 8 + (buffer[1] as u16))
```

Comment:
You removed necessary parentheses, this line should be:
Ok(((buffer[0] as u16) << 8) + (buffer[1] as u16))

but due to the Rust operators precedence rules your version is interpreted as:
Ok((buffer[0] as u16) << (8 + (buffer[1] as u16)))

Even better I propose to change this line to:
Ok((buffer[0] as u16) << 8 | (buffer[1] as u16))

Reply:
Oh wow, I was unaware, thank you!